### ISSUE TITLE
Flexbook content crashes on iPad2.

### ISSUE THREAD

**Issue by cpauya** [learningequality/kolibri#5651]
### Observed behavior
Loading the Flexbook content on iPad2 crashes the web page.  Refer to attached screenshot.

### Expected behavior
Flexbook content should be loaded properly on iPad2 if we are to support the device.

### User-facing consequences
User cannot view Flexbook contents.

### Errors and logs
See attached screenshot.

### Steps to reproduce
Import the `CK-12 Testing` channel using the `nakav-mafak` token.
Navigate to the `CK-12 Testing > HTML5 > Flexbook > Whole Number Subtraction (Flexbook)` content. 

Notice that the iPad2 becomes unresponsive and crashes a few times before showing the `Web Page Crashed`. 

### Context
 * Kolibri version: 0.12.4-beta7
 * Operating system: iOS9.3 on iOS Simulator
 * Browser: Safari 9.3 on iPad 2

![Simulator Screen Shot - iPad 2 - 2019-06-11 at 16 39 15](https://user-images.githubusercontent.com/175580/59261534-3d0d0500-8c70-11e9-8bba-dae7992f4ced.png)

**Comment by indirectlylit** [comment]
do we know if this exists in 0.12.3?

**Comment by cpauya** [comment]
Yes this also exists in 0.12.3.

**Comment by jonboiser** [comment]
I was able to reproduce this on Kolibribeta running (0.12.5). Unfortunately, you can't view the Safari devtools with iOS 9 emulation, but ran it with iOS 10 and got these errors in the console:

The HTML5 app doesn't crash like it does on iOS9, but it does have problems, like it doesn't load an image and I'm unable to scroll on the content.

```
[Error] Refused to load http://kolibribeta.learningequality.org/zipcontent/8021c34324a60bcf0efe94fc6007af58.zip/flex-styles.css because it appears in neither the style-src directive nor the default-src directive of the Content Security Policy.
[Error] Refused to load http://kolibribeta.learningequality.org/zipcontent/8021c34324a60bcf0efe94fc6007af58.zip/146834-1427474629-07-13-5415426436_dc0a8e3ed4_o.jpg because it appears in neither the img-src directive nor the default-src directive of the Content Security Policy.
[Error] Refused to load http://kolibribeta.learningequality.org/zipcontent/8021c34324a60bcf0efe94fc6007af58.zip/201412291419898654271534_6453f42d1398b513177f4dc729b3c9fd-201412291419900375557027.png because it appears in neither the img-src directive nor the default-src directive of the Content Security Policy.
[Error] Refused to load http://kolibribeta.learningequality.org/static/content/hashiframe-9b3b389fbef7cfbe21e5.js because it appears in neither the script-src directive nor the default-src directive of the Content Security Policy.
```

**Comment by jonboiser** [comment]
Something similar seems to be happening on all platforms (tried on newest Safari and Chrome for IOS). Even on Desktop Chrome, I can scroll the HTML5 app for about 5 lines, and then it crashes and becomes unresponsive with these console errors:

```
cookies.min.js:1 Uncaught DOMException: Failed to set the 'cookie' property on 'Document': The document is sandboxed and lacks the 'allow-same-origin' flag.
    at Function.b.set (chrome-extension://hmgpakheknboplhmlicfkkgjipfabmhp/js/cookies.min.js:1:540)
    at Function.b._areEnabled (chrome-extension://hmgpakheknboplhmlicfkkgjipfabmhp/js/cookies.min.js:1:2106)
    at h (chrome-extension://hmgpakheknboplhmlicfkkgjipfabmhp/js/cookies.min.js:1:2189)
    at chrome-extension://hmgpakheknboplhmlicfkkgjipfabmhp/js/cookies.min.js:1:2244
    at chrome-extension://hmgpakheknboplhmlicfkkgjipfabmhp/js/cookies.min.js:1:2465
b.set @ cookies.min.js:1
b._areEnabled @ cookies.min.js:1
h @ cookies.min.js:1
(anonymous) @ cookies.min.js:1
(anonymous) @ cookies.min.js:1
VM2409:2 Refused to load the script 'https://js-agent.newrelic.com/nr-1026.min.js' because it violates the following Content Security Policy directive: "default-src 'self' 'unsafe-inline' 'unsafe-eval' data: blob: http://kolibribeta.learningequality.org". Note that 'script-src-elem' was not explicitly set, so 'default-src' is used as a fallback.

r @ VM2409:2
o @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
t @ hashiframe-9b3b389fbef7cfbe21e5.js:1
o @ hashiframe-9b3b389fbef7cfbe21e5.js:1
load (async)
__nr_require.loader @ VM2409:2
r @ VM2409:2
__nr_require.1 @ VM2409:2
(anonymous) @ VM2409:2
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
t @ hashiframe-9b3b389fbef7cfbe21e5.js:1
d @ hashiframe-9b3b389fbef7cfbe21e5.js:1
h @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
o.handleMessage @ hashiframe-9b3b389fbef7cfbe21e5.js:1
postMessage (async)
Mediator.sendMessage @ mediator.js:50
mainClient_MainClient.__setData @ mainClient.js:64
(anonymous) @ mainClient.js:45
(anonymous) @ mediator.js:24
Mediator.handleMessage @ mediator.js:22
sentryWrapped @ helpers.js:70
postMessage (async)
o.sendMessage @ hashiframe-9b3b389fbef7cfbe21e5.js:1
b @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
r @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
(anonymous) @ hashiframe-9b3b389fbef7cfbe21e5.js:1
jax.js?rev=2.6.0:19 Refused to load the font 'about:blank' because it violates the following Content Security Policy directive: "default-src 'self' 'unsafe-inline' 'unsafe-eval' data: blob: http://kolibribeta.learningequality.org". Note that 'font-src' was not explicitly set, so 'default-src' is used as a fallback.

getDefaultExEm @ jax.js?rev=2.6.0:19
InitializeHTML @ jax.js?rev=2.6.0:19
execute @ MathJax.js:19
WAITEXECUTE @ MathJax.js:19
cb @ MathJax.js:19
execute @ MathJax.js:19
cb @ MathJax.js:19
setTimeout (async)
checkLength @ MathJax.js:19
execute @ MathJax.js:19
cb @ MathJax.js:19
setTimeout (async)
start @ MathJax.js:19
create @ MathJax.js:19
Styles @ MathJax.js:19
Startup @ jax.js?rev=2.6.0:19
execute @ MathJax.js:19
cb @ MathJax.js:19
Process @ MathJax.js:19
call @ MathJax.js:19
WAITEXECUTE @ MathJax.js:19
cb @ MathJax.js:19
Process @ MathJax.js:19
call @ MathJax.js:19
WAITEXECUTE @ MathJax.js:19
cb @ MathJax.js:19
Execute @ MathJax.js:19
loadComplete @ MathJax.js:19
execute @ MathJax.js:19
cb @ MathJax.js:19
Process @ MathJax.js:19
call @ MathJax.js:19
WAITEXECUTE @ MathJax.js:19
cb @ MathJax.js:19
execute @ MathJax.js:19
cb @ MathJax.js:19
setTimeout (async)
checkLength @ MathJax.js:19
execute @ MathJax.js:19
cb @ MathJax.js:19
setTimeout (async)
start @ MathJax.js:19
create @ MathJax.js:19
Styles @ MathJax.js:19
execute @ MathJax.js:19
cb @ MathJax.js:19
Process @ MathJax.js:19
Push @ MathJax.js:19
Init @ MathJax.js:19
Init @ MathJax.js:19
(anonymous) @ MathJax.js:19
(anonymous) @ MathEvents.js?rev=2.6.0:19
(anonymous) @ MathEvents.js?rev=2.6.0:19
kolibribeta.learningequality.org/:1 Access to script at 'http://kolibribeta.learningequality.org/zipcontent/8021c34324a60bcf0efe94fc6007af58.zip/mathjax/jax/output/HTML-CSS/autoload/menclose.js?rev=2.6.0' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
MathJax.js:19 Uncaught RangeError: Maximum call stack size exceeded
    at Object.processError (MathJax.js:19)
    at Object.processOutput (MathJax.js:19)
    at Function.execute (MathJax.js:19)
    at cb (MathJax.js:19)
    at Function.AFTER [as After] (MathJax.js:19)
    at Object.processError (MathJax.js:19)
    at Object.processOutput (MathJax.js:19)
    at Function.execute (MathJax.js:19)
    at cb (MathJax.js:19)
    at Function.AFTER [as After] (MathJax.js:19)
```

**Comment by kollivier** [comment]
The URL http://kolibribeta.learningequality.org/zipcontent/8021c34324a60bcf0efe94fc6007af58.zip/mathjax/jax/output/HTML-CSS/autoload/menclose.js?rev=2.6.0
is not having the CORS headers set on it. Oddly enough another JS file in the same folder has the CORS headers set just fine, so no idea why this URL is special. 

The iOS 10 errors seem really strange. Can you see the network request / response headers for those URLs in the dev console? Particularly I'm wondering what the value of the 'Content-Security-Policy' header on those files is.

**Comment by jonboiser** [comment]
@kollivier In the iOS 10 simulator, I don't see any Content-Security-Policy headers for either request or response for 'menclose.js' or 'fontdata.js' (another file that also seems to consistently give a 404 on some CK-12 resources).

**Comment by kollivier** [comment]
@jonboiser Okay, I think the next step would be to figure out why those headers aren't being added. I don't know when I'll have time to dig into this myself, but the way to test would probably be to download the Testing CK-12 channel locally and then add debugging to the `add_security_headers` decorator in `kolibri/core/content/decorators.py`. Particularly it would be worth checking if and why response is null. @rtibbles Also, why do we try to retrieve the request argument from args / kwargs when it appears to be the first argument passed in?

**Comment by rtibbles** [comment]
> why do we try to retrieve the request argument from args / kwargs when it appears to be the first argument passed in?

Request is always the first argument passed to a view, but if you don't use args/kwargs in the decorator, then if the view takes other arguments, the decorator may blow up, so this is the safest way to define a decorator for a Django view.

**Comment by rtibbles** [comment]
My guess for what is going wrong here is that it is something about the way that Mathjax is loading the files, that is failing to properly account for the crossorigin nature of the request.

Either it is loading it from a script tag, and not adding `crossorigin` property to the tag, or it is loading using something like an XHR request, but is not using XMLHTTPRequest (which we monkey patch in Hashi to automatically add the right CORS information to every request).

**Comment by kollivier** [comment]
Ah, good point, I had expected there to be a lot more errors if the problem was content-related, but if they're loading these specially then that makes sense. Sounds like more fun content hacking to do...!

**Comment by kollivier** [comment]
Actually, I've tried searching for this content in the public CK-12 channel but I can't find it. Is this content only being used for testing? Does anyone know if this content exists in the CK-12 official channel, perhaps under another name? (CC @jamalex) Main issue I'm wondering about is if these issues may be related to this being a testing version of the channel that had outstanding bugs that were fixed in the release version.
